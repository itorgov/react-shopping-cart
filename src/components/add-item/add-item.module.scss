.card {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: center;

  .header {
    color: var(--color-text-dark);
    font-size: 1.8rem;
  }

  .cta-block {
    flex: 1;
    margin-top: 2rem;

    display: flex;
    flex-direction: column;
    justify-content: flex-end;
    align-items: center;

    .price {
      color: var(--color-text-dark);
      font-size: 2rem;
      letter-spacing: 2px;
      margin-bottom: 2rem;
    }

    .added {
      padding: 0.5rem;
      text-transform: uppercase;
      color: var(--color-text-dark);
    }
  }
}
