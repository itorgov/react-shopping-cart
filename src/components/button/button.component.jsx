import React from 'react';
import classNames from "classnames";

import styles from './button.module.scss';

export const Button = ({variant, children: content}) => (
    <button className={classNames([
        styles.button,
        styles[variant || 'primary'],
    ])}
    >
        {content}
    </button>
);
