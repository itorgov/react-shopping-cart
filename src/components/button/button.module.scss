.button {
  font-size: 1.4rem;
  letter-spacing: 3px;
  text-transform: uppercase;
  background: none;
  border: 0;
  border-radius: 10rem;
  padding: 1.5rem 3rem;
  transition: transform 0.2s, box-shadow 0.2s;

  display: flex;
  justify-content: center;
  align-items: center;

  & svg {
    height: 2.5rem;
    width: 2.5rem;
    margin-left: 1rem;
    fill: currentColor;
  }

  &:hover {
    transform: translateY(-2px);
    box-shadow: 0 1rem 2rem rgba(var(--color-secondary-rgb), 0.2);
  }

  &:active {
    transform: translateY(-1px);
    box-shadow: 0 0.5rem 1rem rgba(var(--color-secondary-rgb), 0.2);
  }

  &:focus {
    outline: none;
  }
}

.black {
  background-color: var(--color-text-dark);
  color: var(--color-white);
}
