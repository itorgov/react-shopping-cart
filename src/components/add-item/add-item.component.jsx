import React, {useState} from 'react';
import {useDispatch} from "react-redux";

import {Price} from "../price/price.component";
import {Button} from "../button/button.component";
import {addItem} from "../../redux/cart/cart.actions";

import styles from './add-item.module.scss';

export const AddItem = ({item}) => {
    const [isAdded, setAddedTo] = useState(false);
    const dispatch = useDispatch();

    const addItemToStore = () => {
        dispatch(addItem(item));
        setAddedTo(true);
    }

    return (
        <div className={styles.card}>
            <h2 className={styles.header}>{item.name}</h2>
            <div className={styles['cta-block']}>
                <div className={styles.price}>
                    <Price value={item.price}/>
                </div>
                {
                    isAdded
                        ? <div className={styles.added}>Added</div>
                        : <Button
                            variant='black'
                            small={true}
                            onClick={addItemToStore}
                        >
                            Add to cart
                        </Button>
                }
            </div>
        </div>
    );
};
