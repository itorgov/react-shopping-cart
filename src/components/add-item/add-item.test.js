import React from 'react'
import {mount} from 'enzyme';
import {Provider} from 'react-redux';
import configureStore from 'redux-mock-store'

import {AddItem} from './add-item.component'
import {Button} from "../button/button.component";

let wrapper;
let store;

beforeEach(() => {
    const mockItem = {
        id: 1,
        name: "Test name",
        price: 3900,
    };
    const mockStore = configureStore();
    const initialState = {
        items: [],
        note: '',
    };
    store = mockStore(initialState);
    wrapper = mount(<Provider store={store}><AddItem item={mockItem} /></Provider>);
});

describe('<AddItem />', () => {
    it('expect to render', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('add button has to dispatch ADD_ITEM action', () => {
        wrapper.find(Button).simulate('click');

        const actions = store.getActions();
        expect(actions[0].type).toBe("ADD_ITEM");
    });

    it('add button has to disappear after click', () => {
        expect(wrapper.find(Button).length).toEqual(1);

        wrapper.find(Button).simulate('click');
        wrapper.update();

        expect(wrapper.find(Button).length).toEqual(0);
        expect(wrapper.find('div[className="added"]').text()).toEqual('Added');
    });
});
