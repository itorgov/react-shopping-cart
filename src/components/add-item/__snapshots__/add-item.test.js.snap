// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddItem /> expect to render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <AddItem
    item={
      Object {
        "id": 1,
        "name": "Test name",
        "price": 3900,
      }
    }
  >
    <div
      className="card"
    >
      <h2
        className="header"
      >
        Test name
      </h2>
      <div
        className="cta-block"
      >
        <div
          className="price"
        >
          <Price
            value={3900}
          >
            $39.00
          </Price>
        </div>
        <Button
          onClick={[Function]}
          small={true}
          variant="black"
        >
          <button
            className="button black small"
            onClick={[Function]}
          >
            Add to cart
          </button>
        </Button>
      </div>
    </div>
  </AddItem>
</Provider>
`;
